pilot/pkg/xds/endpoints/endpoint_builder.go

```diff
@@ -795,3 +827,59 @@ func getSubSetLabels(dr *v1alpha3.DestinationRule, subsetName string) labels.Ins
 
 	return nil
 }
+
+// For services that have a waypoint, we want to send to the waypoints rather than the service endpoints.
+// Lookup the
+func (b *EndpointBuilder) findServiceWaypoint(endpointIndex *model.EndpointIndex) ([]*model.IstioEndpoint, bool) {
+	if b.nodeType != model.Router {
+		// Currently only ingress will call waypoints
+		return nil, false
+	}
+	// ...and they must explicitly opt in
+	if b.service.Attributes.Labels["istio.io/ingress-use-waypoint"] != "true" {
+		return nil, false
+	}
+	if b.service.GetAddressForProxy(b.proxy) == constants.UnspecifiedIP {
+		// No VIP, so skip this. Currently, waypoints can only accept VIP traffic
+		return nil, false
+	}
+
+	svcs := b.push.ServicesWithWaypoint(b.service.Attributes.Namespace + "/" + string(b.hostname))
+	if len(svcs) == 0 {
+		// Service isn't captured by a waypoint
+		return nil, false
+	}
+	if len(svcs) > 1 {
+		log.Warnf("unexpected multiple waypoint services for %v", b.clusterName)
+	}
+	svc := svcs[0]
+	waypointClusterName := model.BuildSubsetKey(
+		model.TrafficDirectionOutbound,
+		"",
+		host.Name(svc.WaypointHostname),
```

Comment:
See the comment in [sidecar_interop.go](https://github.com/istio/istio/pull/52985/files#diff-6bd04eb25afa39a37457c09e441ab334f0688142418b40dd786eabb6062dc12b) but currently this may be unset